### PR TITLE
feat: add file reference validator (MSSCI-14579)

### LINKED ISSUE #7: feat: cross-file reference validation for BMB source files and generated output
**Describe your idea**

Add cross-file reference validation to BMad Builder in two layers:

1. **CI validation of BMB's source files** — catch broken references in BMB's 165 source files before they ship to users
2. **Post-generation validation of BMB output** — when BMB creates an agent, workflow, or module, validate that all file references in the generated output resolve to real files

This builds on the cross-file reference validator proposed for BMAD-METHOD (bmad-code-org/BMAD-METHOD#1494), extending it with an `--installed` mode that validates `_bmad/` directories directly.

**Why is this needed?**

### We ran the BMAD-METHOD validator against BMB — here's what it found

| Metric | BMAD-METHOD | BMB |
|---|---|---|
| Source files scanned | 217 | 159 |
| References checked | 483 | 442 |
| Refs per file | 2.2 | 2.8 |
| Broken references found | 5 | 154 |
| Files with issues | 4 | 69 |

The 154 broken references break down as:

| Category | Count | Nature |
|---|---|---|
| Cross-module refs (core/, bmm/, cis/) | 95 | Not bugs — resolve at install time. Validator needs `--skip-external`. |
| **Wrong relative path depth** | **20** | **Genuine bugs** — `../../data/` should be `../data/` in module builder steps. Same class as BMAD-METHOD #1495/#1496. |
| Self-module path mapping | 15 | Validator mapping needs adjustment for BMB's source layout. |
| Template/doc placeholders | 11 | Not bugs — `step-[N]-[name].md` etc. in templates. |
| Example agent references | 5 | Illustrative, not resolvable. |
| Misnamed step reference | 1 | `step-01-validate.md` referenced but actual file is `step-01-load-target.md`. |
| Missing KB CSV files | 3 | Referenced in agent YAML but don't exist in source. |
| Absolute path leaks | 3 | `/Users/` paths in documentation examples. |

The 20 wrong-relative-path bugs affect every step file in the module builder's `steps-b/`, `steps-c/`, `steps-e/`, and `steps-v/` directories — `../../data/` resolves one level too high; the correct path is `../data/`.

In BMAD-METHOD's closed issue history (422 issues, 236 bugs), broken file references and path handling account for **59 bugs (25%)** and 129+ comments. The same reference patterns and bug classes appear in BMB.

### Generated output is not validated for reference correctness

BMB generates agent YAML, workflow markdown, and module configs containing file references (`exec:` paths, `conversational_knowledge` KB paths, `nextStepFile` chains, cross-module refs). BMB's Zod schema validation checks YAML structure but not whether referenced files exist. Users discover broken references at runtime when an LLM tries to load a missing file.

**How should it work?**

### Roadmap

```mermaid
flowchart LR
    subgraph "Step 0 — Prerequisite"
        Z["Reopen & merge<br/>BMAD-METHOD #1493 + PR #1494"]
    end

    subgraph "Step 1 — BMAD-METHOD CI"
        A[validate-file-refs.js] -->|source mode| B["src/ files<br/>~480 refs"]
        B --> C{"Broken?"}
        C -->|Yes| D["⚠️ Warning in build log"]
        C -->|No| E["✓ OK"]
    end

    subgraph "Step 2 — PR Surfacing"
        D --> F["::warning:: annotations<br/>on Files Changed tab"]
        D --> G["PR review comment<br/>(like CodeRabbit)"]
    end

    subgraph "Step 3 — BMB CI"
        A -->|source mode| H["BMB src/ files<br/>~442 refs"]
        H --> I{"Broken?"}
        I -->|Yes| F
        I -->|Yes| G
    end

    subgraph "Step 4 — BMB Output"
        A -->|installed mode| J["User's _bmad/<br/>generated content"]
        J --> K{"Broken?"}
        K -->|Yes| L["🔍 Immediate feedback<br/>to user at creation time"]
    end

    Z --> A
```

Each step is a separate PR. Steps 2–4 build on Step 1 but don't depend on each other.

### Sprint outline

| Step | Work items | Est. |
|------|-----------|------|
| **0. Prerequisite** | Reopen BMAD-METHOD #1493 (closed due to misunderstanding — clarifying comment posted). Merge validator PR #1494 and fix PRs #1497, #1498, #1499. | — |
| **1. BMAD-METHOD CI** | Already complete. Validator runs in `validate` CI job, warning-only (exit 0). | Done |
| **2. PR surfacing** | Add `::warning::` annotations (1 pt), `$GITHUB_STEP_SUMMARY` (1 pt), PR review comment via `workflow_run` (3 pts). | 5 pts |
| **3. BMB CI integration** | Add `--installed` mode to validator (1 pt). Add `--skip-external` for cross-module refs (2 pts). Add `validate:refs` CI step to BMB (2 pts). Surface findings in BMB PRs (2 pts). | 7 pts |
| **4. BMB output validation** | Post-generation validation step in agent builder workflow (2 pts), workflow builder (2 pts), module builder (2 pts). | 6 pts |
| **5. Strict enforcement** | Promote to `--strict` (exit 1) in both repos once existing broken refs are fixed. | 1 pt |
| | **Total** | **19 pts** |

### Unified tool approach

Both CI and output validation use the same validator with a mode flag:
- `--source` (default): maps `{project-root}/_bmad/X` → `src/X` for CI validation
- `--installed`: checks `_bmad/X` directly for installed/generated content
- `--skip-external`: skip cross-module refs that resolve at install time

The reference extraction patterns (regex for `{project-root}/_bmad/`, relative paths, `exec`, `nextStepFile`, etc.) are shared. Only the path resolution differs.

### What this does NOT do

- Does not validate runtime variable interpolation (`{{mustache}}`, `{installed_path}`, etc.)
- Does not modify any generated files — read-only validation
- Does not change existing CI checks or test behavior
- Does not require new dependencies (uses `node:fs`, `node:path`, and `yaml`)

**PR**

We're contributing this work. The BMAD-METHOD validator (PR bmad-code-org/BMAD-METHOD#1494) is ready for review. Once merged, we'll submit PRs for steps 2–4 following the one-fix-per-PR convention.

**Additional context**

- BMAD-METHOD feature request: bmad-code-org/BMAD-METHOD#1493 (closed due to misunderstanding — clarifying comment posted, awaiting response)
- BMAD-METHOD validator PR: bmad-code-org/BMAD-METHOD#1494 (open, passing CI)
- Broken ref fix PRs found by the validator: bmad-code-org/BMAD-METHOD#1497, #1498, #1499
- BMAD-METHOD historical analysis: 422 closed issues classified, 289 unique broken refs tracked across 26 release tags (alpha.0 → Beta.2)

### PR DESCRIPTION
**Pull request by arcaven**
## Summary

- Add `tools/validate-file-refs.mjs` — ESM validator that scans BMB source files for broken cross-file references, absolute path leaks, and stale paths
- Auto-detects module code (`bmb`) from `src/module.yaml` to distinguish internal vs external refs
- Skips external module refs (`core/`, `bmm/`), install-generated files (`config.yaml`, `docs/` KBs), template placeholders (`[N]`, `[name]`), and runtime variables (`{output_folder}`, etc.)
- Inline suppression comments (`<!-- validate-file-refs:ignore -->`) for documentation examples that contain intentionally broken paths
- Data directory trap door: refs in `*/data/` files with `EXAMPLE-` or `invalid-` filename prefixes are auto-skipped
- Reference type metrics: tracks project-root vs relative ref counts, suppressed count, and issue rate percentage
- Add `test/test-validate-file-refs.cjs` with 50 tests covering all 10 acceptance criteria plus suppression and trap door features
- Wire `validate:refs` and `test:refs` npm scripts into the test chain and CI workflow

Fixes #7

## Prior art

This is the bmad-builder port of the Layer 1 file reference validator originally built for BMAD-METHOD:

| PR | Repo | Title | Status |
|----|------|-------|--------|
| [#1494](https://github.com/bmad-code-org/BMAD-METHOD/pull/1494) | BMAD-METHOD | Cross-file reference validator for BMAD source files | ✅ Merged |
| [#1573](https://github.com/bmad-code-org/BMAD-METHOD/pull/1573) | BMAD-METHOD | Extend Layer 1 file-ref validator to scan CSV workflow-file references | ✅ Merged |
| [#1490](https://github.com/bmad-code-org/BMAD-METHOD/pull/1490) | BMAD-METHOD | Cross-file reference validator (predecessor, superseded by #1494) | Closed |
| **#8** | **bmad-builder** | **File reference validator for BMB source tree** | **This PR** |

## Validation pyramid — bmad-builder

```
                ┌─────────────┐
                │   Layer 3   │  Graph Validation
                │             │  Step transitions, reachability
                │             │  (planned)
            ┌───┴─────────────┴───┐
            │      Layer 2        │  Schema & Compilation Validation
            │                     │  Agent schema (Zod), agent compilation,
            │                     │  install component tests
            │                     │  ✅ EXISTS
        ┌───┴─────────────────────┴───┐
        │          Layer 1            │  File Reference Validation
        │                             │  Cross-file refs, path resolution,
        │                             │  absolute path leak detection
        │                             │  ◄── THIS PR
    ┌───┴─────────────────────────────┴───┐
    │              Layer 0                │  Formatting, Linting & Docs
    │                                     │  Prettier, ESLint, markdownlint,
    │                                     │  doc link validation, docs build
    │                                     │  ✅ EXISTS
    └─────────────────────────────────────┘
```

### Implementation status across repos

```mermaid
sequenceDiagram
    participant BMB as bmad-builder
    participant BM as BMAD-METHOD

    Note over BMB: ✅ Layer 0 — Formatting, Linting & Docs<br/>Prettier, ESLint, markdownlint,<br/>doc link validation, docs build

    rect rgba(0, 128, 255, 0.1)
    BMB->>BMB: Layer 1 — File Reference Validator<br/>PR #8 (this PR)
    end

    Note over BMB: ✅ Layer 2 — Schema & Compilation<br/>Agent schema (Zod), agent compilation,<br/>install component tests
    Note over BMB: ⬚ Layer 3 — Graph Validation (planned)

    Note over BMB,BM: ── cross-repo status ──

    Note over BM: ✅ Layer 0 — Prettier, ESLint, markdownlint
    Note over BM: ✅ Layer 1 — File Reference Validator<br/>PR #1494 merged + CSV extension #1573 merged
    Note over BM: ✅ Layer 2 — Agent Schema Validator<br/>PR #774 merged
    Note over BM: ⬚ Layer 2b — Workflow Schema (replanning for workflow*.md)
    Note over BM: ⬚ Layer 3 — Graph Validation (planned)
```

## How it works

```mermaid
flowchart TD
    A[Discover source files in src/<br/>YAML, Markdown, CSV] --> B[Read file content]
    B --> C{File type?}
    C -->|YAML| D[Parse YAML tree<br/>walk maps, sequences, scalars]
    C -->|Markdown| E[Parse frontmatter<br/>+ strip code blocks<br/>+ regex body scan]
    C -->|CSV| F[Parse CSV<br/>extract workflow-file column]
    D --> G[Collect raw references]
    E --> G
    F --> G
    G --> H{Resolvable?}
    H -->|Template placeholder<br/>Runtime variable<br/>Mustache tag| I[Skip]
    H -->|Yes| J{External module?}
    J -->|core/, bmm/, etc.| K["Log [INFO] if --include-external"]
    J -->|Internal bmb/| L{Reference type?}
    L -->|project-root / _bmad| M["mapInstalledToSource()<br/>_bmad/bmb/ → src/"]
    L -->|Relative path| N["path.resolve()<br/>from containing file dir"]
    M --> O{fs.existsSync?}
    N --> O
    O -->|Missing| P["Report [BROKEN]<br/>with file:line"]
    O -->|Found| Q["Report [OK]<br/>in verbose mode"]
    B --> R{Absolute path leak?}
    R -->|"/Users/ /home/ C:\"| S["Report [ABS-PATH]<br/>with line number"]
    P --> T[Summary & exit code]
    Q --> T
    S --> T
    T --> U{--strict flag?}
    U -->|Yes + issues| V[Exit 1]
    U -->|No or clean| W[Exit 0]
```

### Reference types validated

| Pattern | Example | Source |
|---------|---------|-------|
| `{project-root}/_bmad/` paths | `{project-root}/_bmad/bmb/workflows/workflow-create-workflow.md` | Agent YAML, workflow MD |
| Relative paths | `./step-02-analysis.md`, `../data/template.md` | Step MD, workflow MD |
| Frontmatter path keys | `nextStepFile`, `stepTemplate`, `checklist` | Step MD |
| CSV workflow-file column | `_bmad/bmb/workflows/agent/workflow-create-agent.md` | Module help CSV |
| Absolute path leaks | `/Users/dev/project/...`, `C:\Users\...` | Any file |

### What gets skipped

| Category | Examples | Why |
|----------|----------|-----|
| External module refs | `_bmad/core/`, `_bmad/bmm/` | Not in this repo's source tree |
| Install-generated paths | `config.yaml`, `docs/*.md` KBs | Only exist after installation |
| Template placeholders | `[N]`, `[name]`, `[template]` | Filled at runtime |
| Runtime variables | `{output_folder}`, `{bmb_creations_output_folder}` | Resolved during execution |
| Mustache templates | `{{variable_name}}` | Template engine handles |
| Out-of-scope paths | `../../../../core/` (resolves outside `src/`) | Belongs to another module |
| Inline suppression | `<!-- validate-file-refs:ignore -->` on same line | Documentation examples with intentionally broken paths |
| Next-line suppression | `<!-- validate-file-refs:ignore-next-line -->` on prior line | Multi-line suppression |
| Data dir examples | Files in `*/data/` with `EXAMPLE-` or `invalid-` prefix | Example/illustration files by naming convention |

## Current state

26 broken references found against a baseline ratchet of 26. These are pre-existing issues in the source tree — the validator surfaces them without blocking CI (warning-only mode by default, `--strict` for exit 1).

### Reference metrics

| Metric | Count |
|--------|-------|
| References checked | 373 |
| project-root refs | 140 |
| Relative refs | 233 |
| Suppressed (ignored) | 6 |
| Broken references | 26 |
| Absolute path leaks | 0 |
| **Issue rate** | **26 / 373 (7.0%)** |

## CLI flags

| Flag | Behavior |
|------|----------|
| (none) | Warning mode, exit 0 |
| `--strict` | Exit 1 on any issues |
| `--verbose` | Show all checked refs |
| `--include-external` | Show external refs as INFO |

## Suppressing false positives

For documentation files that contain intentionally broken paths (examples, FORBIDDEN patterns tables, etc.), add an inline HTML comment:

```markdown
<!-- Same line -->
| `thisStepFile: './step-XX.md'` | Remove unless referenced <!-- validate-file-refs:ignore --> |

<!-- Previous line -->
<!-- validate-file-refs:ignore-next-line -->
**Frontmatter:** Add `continueFile: './step-01b-continue.md'`
```

For data directory files that serve as examples or illustrations, prefix filenames with `EXAMPLE-` or `invalid-` and they'll be auto-skipped. As archived, legacy, or no-longer-valid content moves into `data/` directories, additional validator handling may emerge as that pattern develops.

## Why is this safe to adopt

*Non-blocking by design.*

The validator runs in **warning mode by default (exit 0)**. Broken references appear in the build log for visibility, but build results are unaffected. No existing CI checks, pre-commit hooks, or npm scripts are modified in behavior. The validator is purely additive.

Every existing CI check continues to enforce exactly as before:

| CI check | Before this PR | After this PR |
|---|---|---|
| Prettier, ESLint, markdownlint | Enforced (exit 1) | Enforced (exit 1) |
| Doc links, docs build | Enforced (exit 1) | Enforced (exit 1) |
| Schema validation, agent tests, install tests | Enforced (exit 1) | Enforced (exit 1) |
| **File ref validation (new)** | *Does not exist* | **Warning only (exit 0)** |

When ready to enforce, one change in `package.json`:

```diff
- "validate:refs": "node tools/validate-file-refs.mjs"
+ "validate:refs": "node tools/validate-file-refs.mjs --strict"
```

The validator only reads files. It makes no changes to disk.

## First-run findings

On first run against `main`, the validator found 32 raw findings (29 broken references + 3 absolute path leaks) across 162 source files (373 references checked, 67 external refs correctly skipped). After analysis, 26 are confirmed real bugs and 6 are false positives (documentation examples). Fix PRs have been filed for 24 of the 26. The 6 false positives are now suppressed with inline ignore comments in this PR.

### All findings

| # | File | Line | Type | Finding | Status | Issue |
|---|------|------|------|---------|--------|-------|
| 1 | `module/steps-c/step-01-load-brief.md` | 7 | BROKEN | `../../templates/agent-spec-template.md` — wrong depth + wrong dir | Valid | [#13](https://github.com/bmad-code-org/bmad-builder/issues/13) |
| 2 | `module/steps-c/step-01-load-brief.md` | 8 | BROKEN | `../../templates/workflow-spec-template.md` — wrong depth | Valid | [#14](https://github.com/bmad-code-org/bmad-builder/issues/14) |
| 3 | `module/steps-c/step-01-load-brief.md` | 9 | BROKEN | `../../data/module-standards.md` — wrong depth | Valid | [#9](https://github.com/bmad-code-org/bmad-builder/issues/9) |
| 4 | `module/steps-c/step-01-load-brief.md` | 10 | BROKEN | `../../data/module-yaml-conventions.md` — wrong depth | Valid | [#10](https://github.com/bmad-code-org/bmad-builder/issues/10) |
| 5 | `module/steps-c/step-02-structure.md` | 6 | BROKEN | `../../data/module-standards.md` — wrong depth | Valid | [#9](https://github.com/bmad-code-org/bmad-builder/issues/9) |
| 6 | `module/steps-c/step-03-config.md` | 6 | BROKEN | `../../data/module-yaml-conventions.md` — wrong depth | Valid | [#10](https://github.com/bmad-code-org/bmad-builder/issues/10) |
| 7 | `module/steps-c/step-04-agents.md` | 6 | BROKEN | `../../templates/agent-spec-template.md` — wrong depth + wrong dir | Valid | [#13](https://github.com/bmad-code-org/bmad-builder/issues/13) |
| 8 | `module/steps-c/step-04-agents.md` | 7 | BROKEN | `../../data/agent-architecture.md` — wrong depth | Valid | [#11](https://github.com/bmad-code-org/bmad-builder/issues/11) |
| 9 | `module/steps-c/step-05-workflows.md` | 6 | BROKEN | `../../templates/workflow-spec-template.md` — wrong depth | Valid | [#14](https://github.com/bmad-code-org/bmad-builder/issues/14) |
| 10 | `module/steps-c/step-07-complete.md` | 8 | BROKEN | `../steps-v/step-01-validate.md` — wrong filename | Valid | [#15](https://github.com/bmad-code-org/bmad-builder/issues/15) |
| 11 | `module/steps-c/step-01b-continue.md` | 5 | BROKEN | `../workflow.md` — wrong filename | Valid | [#18](https://github.com/bmad-code-org/bmad-builder/issues/18) |
| 12 | `module/steps-b/step-13-review.md` | 6 | BROKEN | `../../templates/brief-template.md` — wrong depth | Valid | [#12](https://github.com/bmad-code-org/bmad-builder/issues/12) |
| 13 | `module/steps-b/step-14-finalize.md` | 5 | BROKEN | `../../templates/brief-template.md` — wrong depth | Valid | [#12](https://github.com/bmad-code-org/bmad-builder/issues/12) |
| 14 | `module/steps-e/step-01-load-target.md` | 6 | BROKEN | `../../data/module-standards.md` — wrong depth | Valid | [#9](https://github.com/bmad-code-org/bmad-builder/issues/9) |
| 15 | `module/steps-v/step-02-file-structure.md` | 6 | BROKEN | `../../data/module-standards.md` — wrong depth | Valid | [#9](https://github.com/bmad-code-org/bmad-builder/issues/9) |
| 16 | `module/steps-v/step-03-module-yaml.md` | 6 | BROKEN | `../../data/module-yaml-conventions.md` — wrong depth | Valid | [#10](https://github.com/bmad-code-org/bmad-builder/issues/10) |
| 17 | `module/steps-v/step-04-agent-specs.md` | 6 | BROKEN | `../../templates/agent-spec-template.md` — wrong depth + wrong dir | Valid | [#13](https://github.com/bmad-code-org/bmad-builder/issues/13) |
| 18 | `module/steps-v/step-04-agent-specs.md` | 7 | BROKEN | `../../data/agent-architecture.md` — wrong depth | Valid | [#11](https://github.com/bmad-code-org/bmad-builder/issues/11) |
| 19 | `module/steps-v/step-04-agent-specs.md` | 8 | BROKEN | `{project-root}/…/agent/steps-v/step-01-validate.md` — wrong filename | Valid | [#16](https://github.com/bmad-code-org/bmad-builder/issues/16) |
| 20 | `module/steps-v/step-05-workflow-specs.md` | 6 | BROKEN | `../../templates/workflow-spec-template.md` — wrong depth | Valid | [#14](https://github.com/bmad-code-org/bmad-builder/issues/14) |
| 21 | `module/steps-v/step-08-report.md` | 6 | BROKEN | `{project-root}/…/agent/steps-v/step-01-validate.md` — wrong filename | Valid | [#16](https://github.com/bmad-code-org/bmad-builder/issues/16) |
| 22 | `module/workflow-validate-module.md` | 6 | BROKEN | `./steps-v/step-01-validate.md` — wrong filename | Valid | [#15](https://github.com/bmad-code-org/bmad-builder/issues/15) |
| 23 | `module/workflow-edit-module.md` | 6 | BROKEN | `./steps-e/step-01-assess.md` — wrong filename | Valid | [#19](https://github.com/bmad-code-org/bmad-builder/issues/19) |
| 24 | `agent/steps-e/e-01-load-existing.md` | 7 | BROKEN | `../workflow.md` — wrong filename | Valid | [#17](https://github.com/bmad-code-org/bmad-builder/issues/17) |
| 25 | `workflow/workflow-rework-workflow.md` | 5 | BROKEN | `./steps-r/step-01-assess-rework.md` — missing dir (stub) | Valid | [#20](https://github.com/bmad-code-org/bmad-builder/issues/20) |
| 26 | `workflow/templates/step-1b-template.md` | 20 | BROKEN | `./step-01b-continue.md` — template self-ref | Valid | [#21](https://github.com/bmad-code-org/bmad-builder/issues/21) |
| 27 | `workflow/data/frontmatter-standards.md` | 113 | BROKEN | `./step-XX.md` — example in FORBIDDEN patterns table | Fixed (this PR) | — |
| 28 | `workflow/data/frontmatter-standards.md` | 114 | BROKEN | `./workflow.md` — example in FORBIDDEN patterns table | Fixed (this PR) | — |
| 29 | `workflow/data/step-type-patterns.md` | 82 | BROKEN | `./step-01b-continue.md` — instructional example | Fixed (this PR) | — |
| 30 | `agent/data/agent-validation.md` | 80 | ABS-PATH | `/Users/` — checklist text about what to avoid | Fixed (this PR) | — |
| 31 | `workflow/data/frontmatter-standards.md` | 19 | ABS-PATH | `/Users/user/dev/BMAD-METHOD` — variable example table | Fixed (this PR) | — |
| 32 | `workflow/data/frontmatter-standards.md` | 21 | ABS-PATH | `/Users/user/dev/BMAD-METHOD/output` — variable example table | Fixed (this PR) | — |

### Findings by root cause

| Root cause | Findings | Unique targets | Fix |
|-----------|----------|---------------|-----|
| Wrong path depth (`../../` should be `../`) | 17 | 6 files | [PR #23](https://github.com/bmad-code-org/bmad-builder/pull/23) |
| Wrong validation step filename | 4 | 2 files | [PR #22](https://github.com/bmad-code-org/bmad-builder/pull/22) |
| Wrong workflow/step entry point filename | 3 | 3 files | [PR #24](https://github.com/bmad-code-org/bmad-builder/pull/24) |
| Missing directory / stub feature | 1 | 1 dir | [#20](https://github.com/bmad-code-org/bmad-builder/issues/20) (issue only) |
| Template self-reference | 1 | 1 file | [#21](https://github.com/bmad-code-org/bmad-builder/issues/21) (issue only) |
| Documentation examples (false positive) | 6 | 3 files | Fixed (this PR — inline ignore comments) |
| **Total** | **32** | | |

### Reconciliation

| | Count |
|---|------|
| Validator raw output | 32 (29 broken refs + 3 abs path leaks) |
| False positives (doc examples) | -6 (suppressed with inline ignore comments) |
| **Real findings** | **26** |
| Covered by fix PRs | 24 (across 3 PRs, 11 issues) |
| Issue-only (low confidence) | 2 (2 issues) |

## Test plan

- [x] All 50 tests pass locally (RED → GREEN TDD)
- [x] `npm run validate:refs` runs successfully
- [x] `npm run test:refs` passes
- [x] CI quality workflow includes validate:refs step
- [ ] CI pipeline passes on PR

### Ad hoc: wrong-depth findings verified against simulated install

The 17 wrong-depth findings (`../../` → `../`) are the largest finding category. To confirm these aren't valid at install time (e.g., if the installer adds intermediate directories that would make `../../` correct), we tested paths against both the source tree and a simulated installed output.

**Method:** Ran `bmad-method install` (local dev CLI, non-interactive mode per [docs](https://github.com/bmad-code-org/BMAD-METHOD/blob/main/docs/how-to/non-interactive-installation.md)) to a temp directory, then resolved paths from the installed output.

```bash
node tools/cli/bmad-cli.js install \
  --directory /tmp/bmad-test-install \
  --modules bmb --tools none --user-name TestUser --yes
```

Installed structure — `steps-*`, `data/`, and `templates/` remain siblings under `module/`:

```
_bmad/bmb/workflows/module/
├── data/          ← module-standards.md, agent-architecture.md, etc.
├── templates/     ← brief-template.md, workflow-spec-template.md
├── steps-b/
├── steps-c/       ← step files referencing ../../data/
├── steps-e/
└── steps-v/
```

Path resolution from the installed `steps-c/`:

```bash
cd /tmp/bmad-test-install/_bmad/bmb/workflows/module/steps-c

ls ../data/module-standards.md     # exists ✅
ls ../../data/module-standards.md  # No such file or directory ❌
```

All 17 tested against both source tree and installed output:

| Result | Count | Details |
|--------|-------|---------|
| Confirmed (../../ broken, ../ works) | 14 | Pure depth correction |
| Both broken (wrong depth + wrong dir) | 3 | `templates/agent-spec-template.md` — file lives at `data/agent-spec-template.md` |

All 17 are confirmed bugs. The installer preserves relative directory structure with no path rewriting (`manager.js:762`).

## Release Notes

* **New Features**
  * Added automated validation for cross-file references to detect broken links and absolute path leaks in documentation and configuration files.

* **Tests**
  * Introduced comprehensive test suite with fixtures covering valid references, invalid references, and edge cases such as external references, templates, and runtime variables.

* **Chores**
  * Integrated file reference validation into CI/CD pipeline.
  * Added new npm scripts: `validate:refs` and `test:refs`.
  * Updated documentation with validation markers.

### CODE REVIEW

File: test/test-validate-file-refs.cjs [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+      fail(`export: ${name}`, `missing or not a function (got ${typeof _testing[name]})`);
+    }
+  }
+}
+
+/**
+ * AC3: Path mapping — {project-root}/_bmad/bmb/ → src/
+ */
+async function testPathMapping() {
+  section('AC2: Path mapping');
```
[the remark this answers is not in the document]
Comment (arcaven) [reply test/test-validate-file-refs.cjs]: Good catch, fixed in f5c4018.
